### PR TITLE
Initial script for the model validation tool

### PR DESCRIPTION
**Pull request by ayissi-msft**
This script can do the following when given a hugging face model name:

1. Download the model.
2. Convert the model to ONNX format.
3. Create the model and tokenizer from the downloaded model.
4. Iterate through input texts:
- Embed each input.
- Create a generator.
- Generate output text.

### CODE REVIEW

File: test/python/model_validation/input.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "models": "Qwen/Qwen2-0.5B"
```
Comment (fajin-corp) [inline test/python/model_validation/input.json:2]: >"Qwen/Qwen2-0.5B" [](http://example.com/codeflow?start=14&length=17)

change it to a list to accept multiple models

File: test/python/model_validation/validation_tool.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#             neg_log_likelihood = outputs.loss
+
+#         nlls.append(neg_log_likelihood)
+
+#         prev_end_loc = end_loc
+#         if end_loc == seq_len:
+#             break
+
+#     ppl = torch.exp(torch.stack(nlls).mean())
+#     return
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:55]: remove the unrelated code in this branch. You can move these code to a new branch for perplexity

File: test/python/model_validation/validation_tool.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
+import onnxruntime_genai as og
+import argparse
+import time
+from ...src.python.py.models.builder import create_model
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:4]: from onnxruntime_genai.models.builder import create_model

File: test/python/model_validation/validation_tool.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    # Set the max length to something sensible by default, unless it is specified by the user,
+    # since otherwise it will be set to the entire context length
+
+    search_options['max_length'] = 512
+
+    chat_template = '<|user|>\n{input} <|end|>\n<|assistant|>'
+
+
+    inputs = ["Provide a detailed analysis of the causes and consequences of the French Revolution, including key events, figures, and social changes.",
+              "Explain the process of photosynthesis in plants, detailing the chemical reactions involved, the role of chlorophyll, and the importance of sunlight"]
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:82]: move the inputs to the config.json so users can change it at will

File: test/python/model_validation/validation_tool.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    with open(args.json, 'r') as file:
+        data = json.load(file)
+        args.model = data['models']
+
+    '''
+    1. Download the model 
+    2. Convert to onnx format 
+    3. Create model and tokenizer from the model 
+    '''   
+    model = create_model(args.model, '', './output', 'int4', 'cpu', "./cache")
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:173]: >'./output' [](http://example.com/codeflow?start=41&length=10)

use the "models_outputs" folder in the genai repo as output

Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:173]: >"./cache" [](http://example.com/codeflow?start=68&length=9)

use the "cache_models" folder in the genai repo

File: test/python/model_validation/validation_tool.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+    '''
+    4. Iterate through input texts. In each iteration, embed the input, create generator, generate output text. 
+    5. Log the input text and output text for manual evaluation. 
+    '''
+    validate_model(args, 'output')
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:179]: wrap the code in a loop, so the script can validate multiple models in one run.

**Comment by fajin-corp** [comment]
please try to resolve the auto generated comments as well

File: test/python/model_validation/validation_tool.py (hunk 5 above)
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:173]: >model = create_model(args.model, '', './output', 'int4', 'cpu', "./cache") [](http://example.com/codeflow?start=4&length=74)

add checks after model creation. if the creation failed, log error, and skip current model validation, continue to the next model.

File: test/python/model_validation/validation_tool.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+                
+                value_to_save = tokenizer_stream.decode(new_token)
+
+                complete_text += value_to_save
+
+                print(tokenizer_stream.decode(new_token), end='', flush=True)
+
+                
+                if args.timings: new_tokens.append(new_token)
+        except KeyboardInterrupt:
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:132]: >except KeyboardInterrupt: [](http://example.com/codeflow?start=8&length=25)

if other exceptions are thrown, catch the exception, log, and the validation for current model failed.

File: test/python/model_validation/validation_tool.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    model = create_model(args.model, '', './output', 'int4', 'cpu', "./cache")
+
+    '''
+    4. Iterate through input texts. In each iteration, embed the input, create generator, generate output text. 
+    5. Log the input text and output text for manual evaluation. 
+    '''
+    validate_model(args, 'output')
+
+    '''
+    6. Automatically calculate the perplexity metrics if the model has the corresponding dataset *.
```
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:182]: >Automatically [](http://example.com/codeflow?start=7&length=13)

at the end of the validations, log the summaries of all the models.

File: test/python/model_validation/validation_tool.py (hunk 6 above)
Comment (fajin-corp) [inline test/python/model_validation/validation_tool.py:179]: >validate_model [](http://example.com/codeflow?start=4&length=14)

add checks for validate_model. If exceptions thrown, log the error.

**Comment by ayissi-msft** [comment]
> @ayissi-msft please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree company="Microsoft"

File: test/python/model_validation/validation_tool.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
+import onnxruntime_genai as og
```
Comment (yufenglee) [inline test/python/model_validation/validation_tool.py:1]: you can move the tool to the https://github.com/microsoft/onnxruntime-genai/tree/main/tools/python

File: test/python/model_validation/validation_tool.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    
+    # Set the max length to something sensible by default, unless it is specified by the user,
+    # since otherwise it will be set to the entire context length
+
+    search_options['max_length'] = 512
+
+    chat_template = '<|user|>\n{input} <|end|>\n<|assistant|>'
+
+
+    inputs = ["Provide a detailed analysis of the causes and consequences of the French Revolution, including key events, figures, and social changes.",
```
Comment (yufenglee) [inline test/python/model_validation/validation_tool.py:81]: our eventual goal is to build a tool chain that can get the accuracy/perplexity number given a model (e.g., phi3) and a metric (e.g., MMLU)

File: tools/python/model_validation/validation_tool.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    tokenizer = og.Tokenizer(model)
+    tokenizer_stream = tokenizer.create_stream()
+    if args.verbose: print("Tokenizer created")
+    if args.verbose: print()
+    search_options = {name:getattr(args, name) for name in ['do_sample', 'max_length', 'min_length', 'top_p', 'top_k', 'temperature', 'repetition_penalty'] if name in args}
+    
+    # Set the max length to something sensible by default, unless it is specified by the user,
+    # since otherwise it will be set to the entire context length
+
+    search_options['max_length'] = 512
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:25]: >search_options['max_length'] = 512 [](http://example.com/codeflow?start=4&length=34)

move the default value to the config file

File: tools/python/model_validation/validation_tool.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        create_folder(model_output_dir)
+
+    if not folder_exists(model_cache_dir):
+        create_folder(model_cache_dir)
+
+    output = []
+
+    for model in config["models"]:
+        try:
+            create_model(model, '', model_output_dir+f'/{model}', 'int4', 'cpu', model_cache_dir+f'/{model}')
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:136]: >model [](http://example.com/codeflow?start=57&length=5)

if the `model` has '/' or '\', model path parsing may have troubles. better convert '/' and '\' to '_'.

File: tools/python/model_validation/validation_tool.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+            generation_successful = validate_model(config, model_output_dir)
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:137]: >model_output_dir [](http://example.com/codeflow?start=59&length=16)

should it be `model_output_dir+f'/{model}'`?

File: tools/python/model_validation/validation_tool.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    parser.add_argument('-j', '--json', type=str, required=True, help='Path to the JSON file containing the arguments')
+
+    args = parser.parse_args()
+
+    with open(args.json, 'r') as file:
+        config = json.load(file)
+
+    # Check and see if the folder exists, if not create the folder
+    model_output_dir = "../../../models_outputs/"
+    model_cache_dir = "../../../cache_models"
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:124]: move these two variables to config.json as well

File: tools/python/model_validation/validation_tool.py (hunk 12 above)
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:136]: >model_output_dir+f'/{model}' [](http://example.com/codeflow?start=36&length=28)

nit: f'{model_output_dir}/{model}'

File: tools/python/model_validation/validation_tool.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+    if not folder_exists(model_output_dir):
+        create_folder(model_output_dir)
+
```
Comment (kunal-vaishnavi) [inline tools/python/model_validation/validation_tool.py:128]: Using `os.makedirs(folder_path, exist_ok=True)` may be a better option here. If `folder_path` exists, then `folder_path` isn't created and the existing `folder_path` is used. If `folder_path` does not exist, then `folder_path` is created. Here is the [documentation](https://docs.python.org/3/library/os.html#os.makedirs) for more info.

File: tools/python/model_validation/validation_config.json [hunk 16]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (yufenglee) [inline tools/python/model_validation/validation_config.json:1]: we are building a general tool. This model specific config should not appear in the folder

File: tools/python/model_validation/validation_tool.py (hunk 14 above)
Comment (yufenglee) [inline tools/python/model_validation/validation_tool.py:124]: we may need users to provide the option.

File: tools/python/model_validation/validation_tool.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+    if config["verbose"]: print()
+
+    search_option_keys = [
+        'do_sample', 'max_length', 'min_length', 'top_p', 'top_k', 
+        'temperature', 'repetition_penalty'
+    ]   
+
+    search_options = {key: config[key] for key in search_option_keys} 
+
+    chat_template = '<|user|>\n{input} <|end|>\n<|assistant|>'
```
Comment (yufenglee) [inline tools/python/model_validation/validation_tool.py:34]: this is phi3 specific template. it should not be here.

File: tools/python/model_validation/validation_tool.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        generation_successful = True
+
+        try:
+            while not generator.is_done():
+                generator.compute_logits()
+                generator.generate_next_token()
+                if config["timings"]:
+                    if first:
+                        first_token_timestamp = time.time()
+                        first = False
```
Comment (yufenglee) [inline tools/python/model_validation/validation_tool.py:71]: we don't care about the timings in the validation/accuracy tool chain.

File: tools/python/model_validation/validation_config.json [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    "models": [
+    ],
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_config.json:3]:
```
{ "model_name": "qwen", "chat_template":"xxxxx"}
```

File: tools/python/model_validation/README.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+### Usage - Build the Model
+This step creates optimized and quantized ONNX models that run with ONNX Runtime GenAI.
+
+1. In the validation_config.json file, enter the supported Hugging Face model name. Models can be found here.
+2. Include the path to the output folder, precision, and execution provider.
+
+Once the model is built, you can find it in path_to_output_folder/{model_name}. This should include the ONNX model data and tokenizer.
+
+### Run the Model Validation Script
```
Comment (fajin-corp) [inline tools/python/model_validation/README.md:29]: the validation tool automatically run the model builder. the narrative here makes users think this is a separate step.

File: tools/python/model_validation/validation_config.json [hunk 21]
```diff
@@ -0,0 +1,23 @@
+{
+    "models": [
+        "Qwen/Qwen2-0.5B"
+    ],
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_config.json:4]:
```
{ "models": [{"name":"Qwen/Qwen2-0.5B", "chat_template": "xxxx"}]}
```

File: tools/python/model_validation/validation_tool.py [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+    os.makedirs(config["output_directory"], exist_ok=True)
+    os.makedirs(config["cache_directory"], exist_ok=True)
+
+    output = []
+
+    validation_complete = False
+    e = None
+    exception = False
+
+    for model in config["models"]:
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:108]: >model [](http://example.com/codeflow?start=8&length=5)

`model_dict`
`model_dict['name'], model_dict['chat_template']`

File: tools/python/model_validation/README.md [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+### Current Support
+* Gemma
+* Llama 
+* Mistral
+* Phi
+* Qwen
+
+### Usage 
+1. In the validation_config.json file, enter the supported Hugging Face model name. Models can be found here.
```
Comment (fajin-corp) [inline tools/python/model_validation/README.md:22]: add guidance of how to get chat template for a model

File: tools/python/model_validation/validation_config.json [hunk 24]
```diff
@@ -0,0 +1,24 @@
+{
+    "models": [
+        {
+            "name": "",
+            "chat_template": ""
+        }
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_config.json:6]: add qwen, llama, phi3, gemma, mistral

File: tools/python/model_validation/validation_config.json [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+    ],
+    "inputs": [
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_config.json:8]: >inputs": [ [](http://example.com/codeflow?start=5&length=10)

give default inputs

File: tools/python/model_validation/validation_config.json (hunk 16 above)
Comment (fajin-corp) [reply tools/python/model_validation/validation_config.json:1]: updated the configs to include only gen-ai configs + some defaults to facilitate future use cases.

File: tools/python/model_validation/validation_tool.py [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+    validation_complete = False
+    e = None
+    exception = False
+
+    for model_dict in args["models"]:
+
+        print(f"We are validating {model_dict['name']}")
+        adjusted_model = model_dict["name"].replace("/", "_")
+
+        output_path = args["output_directory"] + f'/{adjusted_model}'
```
Comment (fajin-corp) [inline tools/python/model_validation/validation_tool.py:100]: >args["output_directory"] + f'/{adjusted_model}' [](http://example.com/codeflow?start=22&length=47)

os.path

**Review by fajin-corp — APPROVED** [review]
:shipit:

File: tools/python/model_validation/validation_config.json [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+            "chat_template": "<s>[INST]<<SYS>>\n{input}<</SYS>>[INST]"
+        }
+    ],
+    "inputs": [
+        "Provide a detailed analysis as to why the University of Southern California is better than the University of California, Los Angeles."
+    ],
+    "output_directory": "",
+    "cache_directory": "",
+    "precision": "",
+    "executive_provider": "",
```
Comment (kunal-vaishnavi) [inline tools/python/model_validation/validation_config.json:30]:
```suggestion
    "execution_provider": "",
```

File: tools/python/model_validation/validation_tool.py [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+        cache_path = args["cache_directory"] + f'/{adjusted_model}'
+
+        try:
+            create_model(model_dict["name"], '', output_path, args["precision"], args["executive_provider"], cache_path)
```
Comment (kunal-vaishnavi) [inline tools/python/model_validation/validation_tool.py:105]:
```suggestion
            create_model(model_dict["name"], '', output_path, args["precision"], args["execution_provider"], cache_path)
```

**Review by fajin-corp — APPROVED** [review]
